src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -132,10 +146,25 @@ private static TypeSymbol InferReturnType(BoundBlock block, Binder binder, bool
 
             // Async:
 
+            NamedTypeSymbol taskType = null;
+            var delegateReturnType = delegateType?.GetDelegateType()?.DelegateInvokeMethod?.ReturnType as NamedTypeSymbol;
```

Comment:
Unfortunately `DelegateReturnType` is defined on `UnboundLambdaState` not `BoundLambda`.